### ISSUE TITLE
OSM - Badly rendered coast line (tile artifacts)

### ISSUE THREAD

**Issue by aerostitch** [aerostitch/testnavit#93]
*Issue migrated from trac ticket # 854*

**component:** mapdrivers/OSM | **priority:** major | **keywords:** rendering, layout, map, binfile

#### 2011-05-09 14:22:54: pini@debian.org created the issue

___
> Hi,
> 
> Depending on the zoom level the coastline is very badly rendered with big squares when using OSM maps. See the attached screenshot which was obtained using a recent planet.bin. It shows the french town Brest.
> [[Image(Capture-Navit.png​)]]

**Comment by aerostitch** [comment]
#### 2011-05-09 14:23:41: pini@debian.org uploaded file [`Capture-Navit.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Capture-Navit.png) (72.3 KiB)


![Capture-Navit.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Capture-Navit.png)

**Comment by aerostitch** [comment]
#### 2013-06-26 13:50:26: usul commented


___
> That shows 2 issues:
> * empty vectortiles are filled with the default earth colour but should become filled with water instead
> * at some areas, the inner coastline is broken (multipolygone issue) and so it's a rectangle earth surface towards the sea

**Comment by aerostitch** [comment]
#### 2013-07-06 11:24:17: usul

**Comment by aerostitch** [comment]
#### 2013-07-06 11:24:17: usul commented


___
> Is essential, as users might think that the map is somewhat corrupted. -> scheduled for hotfix

**Comment by aerostitch** [comment]
#### 2013-10-14 07:37:39: usul changed severity from ** to *complex*

**Comment by aerostitch** [comment]
#### 2013-10-14 07:37:39: usul edited the issue description

**Comment by aerostitch** [comment]
#### 2013-10-14 07:38:07: usul changed title from *OSM - Badly rendered coast line* to *OSM - Badly rendered coast line (tile artifacts)*

**Comment by aerostitch** [comment]
#### 2015-03-02 08:42:01: @jandegr commented


___
> If you zoom in, things gradually improve and the coastline becomes perfect when zoomed in more.
> This points to a tiledepth issue and setting a maximal tiledepth for poly_water_tiled confirms it.
> I had good results with the attached patch.
> Anyone pls. comment on this or test the patch since this issue really needs to be fixed.
>

**Comment by aerostitch** [comment]
#### 2015-03-02 08:43:02: @jandegr uploaded file [`coast.diff`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/coast.diff) (0.4 KiB)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:43:02: @pgrandin uploaded file [`level-2.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-2.png) (23.3 KiB)


![level-2.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-2.png)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:43:14: @pgrandin uploaded file [`level-3.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-3.png) (8.1 KiB)


![level-3.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-3.png)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:43:27: @pgrandin uploaded file [`level-4.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-4.png) (7.2 KiB)


![level-4.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-4.png)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:43:38: @pgrandin uploaded file [`level-5.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-5.png) (5.0 KiB)


![level-5.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-5.png)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:43:47: @pgrandin uploaded file [`level-6.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-6.png) (2.9 KiB)


![level-6.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/level-6.png)

**Comment by aerostitch** [comment]
#### 2015-03-03 09:54:09: @pgrandin changed status from *new* to *closed*

**Comment by aerostitch** [comment]
#### 2015-03-03 09:54:09: @pgrandin changed resolution from ** to *fixed*

**Comment by aerostitch** [comment]
#### 2015-03-03 09:54:09: @pgrandin commented


___
> I did some tests, and I can confirm that Jan's patch address this. One line, beautiful!
> 
> This is the same coast, viewed from different zoom levels:
> 
> [[Image(level-2.png)]]
> 
> [[Image(level-3.png)]]
> 
> [[Image(level-4.png)]]
> 
> [[Image(level-5.png)]]
> 
> [[Image(level-6.png)]]
> 
> 
> Here's the script to generate the images :
> ```
> prefix="pre854"
> for i in `seq 1 10`; do
>         dbus-send  --print-reply --session --dest=org.navit_project.navit /org/navit_project/navit/default_navit org.navit_project.navit.navit.zoom int32:-2
>         sleep 1;
>         import -window root $prefix-zoom-${i}.png
> done
> ```
> 
> (run once with prefix pre854, then change the map and change the prefix )
> 
> Here is the script to build the images:
> ```
> for i in `seq 1 9`; do
> 
>         convert pre854-zoom-${i}.png -crop 40%x+0+0 pre-cropped-${i}.png
>         convert post854-zoom-${i}.png -crop 40%x+0+0 post-cropped-${i}.png
>         montage pre-cropped-${i}.png post-cropped-${i}.png -tile 2x1 -geometry +0+0 level-${i}.png;
> done
> ```
> 
> Applied in [6017](http://sourceforge.net/p/navit/code/6017/). Good job Jan!
>

**Comment by aerostitch** [comment]
#### 2015-03-04 01:31:45: tryagain changed status from *closed* to *reopened*

**Comment by aerostitch** [comment]
#### 2015-03-04 01:31:45: tryagain changed resolution from *fixed* to **

**Comment by aerostitch** [comment]
#### 2015-03-04 01:31:45: tryagain commented


___
> Hi!
> 
> Do you have whole planet data converted with current maptool version?
> 
> I'm afraid there are hard performance consequences after applying the patch. You have pulled coastline data to tiles of order=6, which have 40000/2^6^ = 625 km side length.
> 
> This will affect all map operations, because you'll have always process coastline data from square 625x625km once your standing point is within 625km from coastline or closer. This might be not a very big impact on powerful devices with lots of RAM, which unzip the tile data only once and then keep it in memory cache. Users with low end devices with cache size reduced, lower cpu and disk io performance, may have dramatic consequences.
> 
> I would like to have tilesdir.tmp file from processed planet with this patch applied to do some analysis, and planet.bin file itself.
> 
> For now, reopening this ticket and marking the new feature as experimental in [6018](http://sourceforge.net/p/navit/code/6018/).
> 
> I guess we would have to do some simplification at maptool time to have smooth coastline without affecting enduser performance. But let us check.
> 
> tryagain.
> 
>

**Comment by aerostitch** [comment]
#### 2015-03-04 14:19:15: @pgrandin commented


___
> That's a valid point Tryagain.
> 
> I did process only a small subset of the planet for this test. Processing time was about the same and I did not see an impact on my test machine, but it is a regular computer, not a low power device.

**Comment by aerostitch** [comment]
#### 2015-03-05 06:50:19: @jandegr uploaded file [`Bordeaux1.JPG`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Bordeaux1.JPG) (30.2 KiB)


![Bordeaux1.JPG](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Bordeaux1.JPG)

**Comment by aerostitch** [comment]
#### 2015-03-05 06:53:26: @jandegr uploaded file [`Bordeaux1.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Bordeaux1.png) (117.4 KiB)


![Bordeaux1.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/Bordeaux1.png)

**Comment by aerostitch** [comment]
#### 2015-03-05 06:54:15: @jandegr commented


___
> Time for a little experiment :
> 
> download (approx.) this area from the mapserver
> 
> [[Image(https://dl.dropboxusercontent.com/u/93775123/Navit/Bordeaux2.JPG)]]
> 
> You see right away that the size of the map is very big for such a small area.
> On other occasions I tried to download the map of a very tiny island, surrounded by
> lots of water and did not get the map below 75MB (from the mapserver).
> 
> Once installed in Navit, zoom in and out a little untill you notice that as a treat the mapserver gave you (chunks of) country borders from all over the globe.
> Maybe after tweaking navit.xml more global-wide content could appear.
> 
> [[Image(https://dl.dropboxusercontent.com/u/93775123/Navit/Bordeaux1.JPG)]]
> 
> Now anyone can try to convince me that some usefull coastlines are worse for ancient hardware than this kind of *%µ# thousands of kilometers of country borders !
> 
> It is in fact exactly the same problem as with the coastlines : randomness all over.
> There is a way to limit tiledepth but to solve both coastlines and the global-wide crap we need to impose an upper tilelevel for objects.
> Some signatures of functions in maptool suggest it is possible since they have a max and min parameter, but the code itself seems to ignore the min parameter.
> 
> If you zoom in a little you will notice the same happens with landuse polygons, roads, ...
> Needless to say that the level=6 I used at first is open for discussion, but to avoid those artefacts and randomness I had to use 6 awaiting a solution to impose an upper limit.
> 
> And something not related to the issue but to the bordeaux area : can someone help me figure out why the bordeaux and bassin d'Arcachon area are flooded (maps from the mapserver and selfbuilt maps just the same) ?
> thx.

**Comment by aerostitch** [comment]
#### 2015-03-05 06:54:15: @jandegr

**Comment by aerostitch** [comment]
#### 2015-03-05 06:54:15: @jandegr

**Comment by aerostitch** [comment]
#### 2015-03-05 06:54:15: @jandegr

**Comment by aerostitch** [comment]
#### 2015-03-05 06:54:15: @jandegr

**Comment by aerostitch** [comment]
#### 2015-03-05 13:34:59: tryagain commented


___
> Replying to [comment:10 http://wiki.navit-project.org/index.php/user:jandegr]:
> 
> > lots of water and did not get the map below 75MB (from the mapserver).
> 
> I think most of this data is zip file directory. It's stored uncompressed. We have really lots of tiles, each tile being a file inside binfile. I hope navit does not reread this data on each map access, but it have to read it at least once. There's lots of data (tile names) which is actually never used by navit itself. If a tile is outside of your area of interest, it's still included in the zip file, but is marked as having zero size.
> 
> Binfile structure is described [in wiki](http://wiki.navit-project.org/index.php/Binfile)
> 
> 
> > Once installed in Navit, zoom in and out a little untill you notice that as a treat the mapserver gave you (chunks of) country borders from all over the globe.
> > Maybe after tweaking navit.xml more global-wide content could appear.
> 
> Yes, it may appear. Actually, any item which crosses tile boundary, will be shifted uplevel, until it does not cross the tile boundary. Items which cross equator, for example, are put into the top level tile. We could workaround it by splitting such items by tile boundary. It's already done for coastlines.
> 
> 
> > Now anyone can try to convince me that some usefull coastlines are worse for ancient hardware than this kind of *%µ# thousands of kilometers of country borders !
> 
> Then contribute code to split and push these country borders deeper ;)
> 
> 
> > It is in fact exactly the same problem as with the coastlines : randomness all over.
> 
> I see your point, but there's no randomness in maptool at all (it does not access /dev/random ;)
> 
> 
> > There is a way to limit tiledepth but to solve both coastlines and the global-wide crap we need to impose an upper tilelevel for objects.
> > Some signatures of functions in maptool suggest it is possible since they have a max and min parameter, but the code itself seems to ignore the min parameter.
> 
> Yes. But we shouldn't just move stuff to upper level. Big objects should be simplified beforehand. And we should take measures to draw only one version of each object at a given time.
> 
> Another way would be add support for prerendered tiles, which could be used to display world map, having vector and prerendered maps swithced at some zoom level.
> 
> > If you zoom in a little you will notice the same happens with landuse polygons, roads, ...
> > Needless to say that the level=6 I used at first is open for discussion, but to avoid those artefacts and randomness I had to use 6 awaiting a solution to impose an upper limit.
> 
> If you make all coastlines accessible at all zoom levels, prepare to process 360+Mb of unzipped coastline data on each map redraw. That's around 45mln points which should be scaled to fit the screen, and 45mln lines drawn.
> 
> 
> > And something not related to the issue but to the bordeaux area : can someone help me figure out why the bordeaux and bassin d'Arcachon area are flooded (maps from the mapserver and selfbuilt maps just the same) ?
> See #745 and [wiki](http://wiki.navit-project.org/index.php/Troubleshooting#Incomplete_Coastlines_within_OSM_Data)
> 
> tryagain.
>

**Comment by aerostitch** [comment]
#### 2015-03-19 12:25:02: tryagain commented


___
> Experimental maptool has produced much smoother coastline, but we now have 92 order=6 tiles above 1Mb, biggest over 8Mb.
> 
> Beforehand we had only 9 order=6 tiles, biggest ones of 2Mb size.
> 
> So applied [6026](http://sourceforge.net/p/navit/code/6026/) to move coastline to order=9.

**Comment by aerostitch** [comment]
#### 2015-03-19 12:29:54: tryagain uploaded file [`GB-DE_order6coastline.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/GB-DE_order6coastline.png) (41.3 KiB)

View of region with coastline moved to order=6
![GB-DE_order6coastline.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/GB-DE_order6coastline.png)

**Comment by aerostitch** [comment]
#### 2015-03-19 12:30:28: tryagain uploaded file [`GB-DE_defaulcoastline.png`](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/GB-DE_defaulcoastline.png) (28.2 KiB)

View of region with default coastline tiling
![GB-DE_defaulcoastline.png](https://raw.githubusercontent.com/aerostitch/navit-attachments/master/854/GB-DE_defaulcoastline.png)

**Comment by aerostitch** [comment]
#### 2017-10-03 04:03:53: wiki.navit-project.org/index.php/user:naggety commented


___
> I did some tests with the experimental feature, even seting order # 1 (max1).
> 
> The RAM memory usage was up to around 60 - 70MB.
